File: wasmbinding/bindings/msg.go
package bindings

import (
	sdk "github.com/cosmos/cosmos-sdk/types"
)

type QuasarMsg struct {
	/// Trigger test scenario
	TestScenario *TestScenario `json:"test_scenario,omitempty"`

	/// Contracts can send tokens
	SendToken *SendToken `json:"send_token,omitempty"`

	// Contracts can register interchain accounts
	RegisterICAOnZone *RegisterICAOnZone `json:"register_ica_on_zone,omitempty"`

	/// Contracts can transmit JoinPool Messages over IBC
	OsmosisJoinPool *OsmosisJoinPool `json:"join_pool,omitempty"`

	/// Contracts can transmit ExitPool Messages over IBC
	OsmosisExitPool *OsmosisExitPool `json:"exit_pool,omitempty"`

	/// Contracts can transmit LockTokens Messages over IBC
	OsmosisLockTokens *OsmosisLockTokens `json:"lock_tokens,omitempty"`

	/// Contracts can start the unbonding process over IBC
	OsmosisBeginUnlocking *OsmosisBeginUnlocking `json:"begin_unlocking,omitempty"`

	// Contracts can transmit JoinSwapExternAmountIn Messages over IBC
	OsmosisJoinSwapExternAmountIn *OsmosisJoinSwapExternAmountIn `json:"join_swap_extern_amount_in,omitempty"`

	// Contracts can transmit ExitSwapExternAmountOut Messages over IBC
	OsmosisExitSwapExternAmountOut *OsmosisExitSwapExternAmountOut `json:"exit_swap_extern_amount_out,omitempty"`
}

type TestScenario struct {
	Scenario string `json:"scenario"`
}

type RegisterICAOnZone struct {
	ZoneID string `json:"zone_id"`
}

type SendToken struct {
	DestinationLocalZoneID string   `json:"destination_local_zone_id"`
	Receiver               string   `json:"receiver"`
	Coin                   sdk.Coin `json:"coin"`
}

type OsmosisJoinPool struct {
	ConnectionID     string     `json:"connection_id"`
	TimeoutTimestamp uint64     `json:"timeout_timestamp"`
	PoolID           uint64     `json:"pool_id"`
	ShareOutAmount   int64      `json:"share_out_amount"`
	TokenInMaxs      []sdk.Coin `json:"token_in_maxs"`
}

type OsmosisExitPool struct {
	ConnectionID     string     `json:"connection_id"`
	TimeoutTimestamp uint64     `json:"timeout_timestamp"`
	PoolID           uint64     `json:"pool_id"`
	ShareInAmount    int64      `json:"share_in_amount"`
	TokenOutMins     []sdk.Coin `json:"token_out_mins"`
}

type OsmosisLockTokens struct {
	ConnectionID     string     `json:"connection_id"`
	TimeoutTimestamp uint64     `json:"timeout_timestamp"`
	Duration         uint64     `json:"duration"`
	Coins            []sdk.Coin `json:"coins"`
}

type OsmosisBeginUnlocking struct {
	ConnectionID     string     `json:"connection_id"`
	TimeoutTimestamp uint64     `json:"timeout_timestamp"`
	ID               uint64     `json:"id"`
	Coins            []sdk.Coin `json:"coins"`
}

type OsmosisJoinSwapExternAmountIn struct {
	ConnectionID      string   `json:"connection_id"`
	TimeoutTimestamp  uint64   `json:"timeout_timestamp"`
	PoolID            uint64   `json:"pool_id"`
	ShareOutMinAmount int64    `json:"share_out_min_amount"`
	TokenIn           sdk.Coin `json:"token_in"`
}

type OsmosisExitSwapExternAmountOut struct {
	ConnectionID     string   `json:"connection_id"`
	TimeoutTimestamp uint64   `json:"timeout_timestamp"`
	PoolID           uint64   `json:"pool_id"`
	ShareInAmount    int64    `json:"share_in_amount"`
	TokenOutMins     sdk.Coin `json:"token_out_mins"`
}

File: smart-contracts/contracts/intergamm-bindings-test/src/msg.rs
use cosmwasm_std::{Coin, Uint64};
use intergamm_bindings::msg::IntergammMsg;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
pub struct InstantiateMsg {
    pub callback_address: String,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ExecuteMsg {
    SendToken {
        destination_local_zone_id: String,
        receiver: String,
        coin: Coin,
    },
    SendTokenIbc {
        /// exisiting channel to send the tokens over
        channel_id: String,
        /// address on the remote chain to receive these tokens
        to_address: String,
        /// packet data only supports one coin
        /// https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/ibc/applications/transfer/v1/transfer.proto#L11-L20
        amount: Coin,
        // when packet times out, measured on remote chain, for now we hardcode the timeout
        // timeout: IbcTimeout,
    },
    Deposit {},
    RegisterIcaOnZone {
        zone_id: String,
    },
    JoinPool {
        connection_id: String,
        timeout_timestamp: Uint64,
        pool_id: Uint64,
        share_out_amount: i64,
        token_in_maxs: Vec<Coin>,
    },
    ExitPool {
        connection_id: String,
        timeout_timestamp: Uint64,
        pool_id: Uint64,
        share_in_amount: i64,
        token_out_mins: Vec<Coin>,
    },
    LockTokens {
        connection_id: String,
        timeout_timestamp: Uint64,
        duration: Uint64,
        coins: Vec<Coin>,
    },
    JoinSwapExternAmountIn {
        connection_id: String,
        pool_id: Uint64,
        share_out_min_amount: i64,
        token_in: Coin,
    },
    ExitSwapExternAmountOut {
        connection_id: String,
        timeout_timestamp: Uint64,
        pool_id: Uint64,
        share_in_amount: i64,
        token_out_mins: Coin,
    },
    BeginUnlocking {
        connection_id: String,
        timeout_timestamp: Uint64,
        id: Uint64,
        coins: Vec<Coin>,
    },
    TestIcaScenario {},
    Ack {
        sequence_number: u64,
        error: Option<String>,
        response: Option<intergamm_bindings::msg::AckResponse>,
    },
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum QueryMsg {
    PendingAcks {},
    Acks {},
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub struct AcksResponse {
    pub acks: Vec<(u64, intergamm_bindings::msg::AckValue)>,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub struct PendingAcksResponse {
    pub pending: Vec<(u64, IntergammMsg)>,
}
